### PR TITLE
MESH-1408/Refactor canTransfer to return all Results

### PR DESCRIPTION
**Pull request by CJP10**
### new features

- new asset rpc `can_transfer_granular`

### CODE REVIEW

File: primitives/src/asset.rs [hunk 1]
```diff
@@ -70,3 +73,35 @@ impl Default for AssetType {
...
+    /// Sender had an insufficient balance
+    pub sender_insufficient_balance: bool,
+    /// Portfolio error
+    pub portfolio_error: bool,
+    /// Asset is frozen
+    pub asset_frozen: bool,
+    /// Statistics check failed
+    pub statistics_failure: bool,
+    /// Compliance check failed
+    pub compliance_failure: bool,
```
Comment (maxsam4) [inline primitives/src/asset.rs:104]: We need these to be more granular. See the can transfer in compliance pallet. Result of that call should be included here basically.

File: primitives/src/asset.rs [hunk 2]
```diff
@@ -70,3 +73,35 @@ impl Default for AssetType {
...
+    /// Sender had a custodian error
+    pub sender_custodian_error: bool,
+    /// Sender had an insufficient balance
+    pub sender_insufficient_balance: bool,
+    /// Portfolio error
+    pub portfolio_error: bool,
+    /// Asset is frozen
+    pub asset_frozen: bool,
+    /// Statistics check failed
+    pub statistics_failure: bool,
```
Comment (maxsam4) [inline primitives/src/asset.rs:102]: We need these to be more granular. i.e. It should list all transfer managers in statistics module and if they passed/failed.

**Review by maxsam4 — CHANGES_REQUESTED** [review]

File: pallets/statistics/src/lib.rs [hunk 3]
```diff
@@ -253,6 +239,43 @@ impl<T: Trait> Module<T> {
...
+                    TransferManager::PercentageTransferManager(max_percentage) => Self::ensure_ptm(
+                        ticker,
+                        receiver,
+                        value,
+                        receiver_balance,
+                        total_supply,
+                        max_percentage,
+                    )
+                    .is_err(),
+                },
```
Comment (vkomenda) [inline pallets/statistics/src/lib.rs:274]: Matching on `&tm` first should allow to remove cloning from `tm`.

File: pallets/asset/src/lib.rs [hunk 4]
```diff
@@ -1609,39 +1609,15 @@ impl<T: Trait> Module<T> {
             return Ok(ERC1400_TRANSFERS_HALTED);
         }
 
-        if !Identity::<T>::verify_iu_claims_for_transfer(
-            *ticker,
-            to_portfolio.did,
-            from_portfolio.did,
-        ) {
+        if Self::missing_scope_claim(ticker, to_portfolio.did, from_portfolio.did) {
```
Comment (Centril) [inline pallets/asset/src/lib.rs:1612]: Function is always called with `portfolio.did`, so let's take the portfolio ids instead

File: pallets/runtime/common/src/runtime.rs [hunk 5]
```diff
@@ -833,6 +832,19 @@ macro_rules! runtime_apis {
...
+
+                #[inline]
+                fn can_transfer_granular(
+                    _sender: polymesh_primitives::AccountId,
+                    from_custodian: Option<IdentityId>,
+                    from_portfolio: PortfolioId,
+                    to_custodian: Option<IdentityId>,
+                    to_portfolio: PortfolioId,
+                    ticker: &Ticker,
+                    value: Balance) -> polymesh_primitives::asset::GranularCanTransferResult
```
Comment (Centril) [inline pallets/runtime/common/src/runtime.rs:844]:
```suggestion
                    value: Balance
                ) -> polymesh_primitives::asset::GranularCanTransferResult
```

File: primitives/src/asset.rs [hunk 6]
```diff
@@ -70,3 +74,35 @@ impl Default for AssetType {
     Decode, Encode, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord, Default, VecU8StrongTyped,
 )]
 pub struct FundingRoundName(pub Vec<u8>);
+
+/// Result of a granular can transfer
```
Comment (Centril) [inline primitives/src/asset.rs:78]: Nit: add punctuation in doc comments.

File: rpc/runtime-api/src/asset.rs [hunk 7]
```diff
@@ -61,5 +61,16 @@ sp_api::decl_runtime_apis! {
             ticker: &Ticker,
             value: Balance
         ) -> CanTransferResult;
+
+         /// TODO
```
Comment (fmiguelgarcia) [inline rpc/runtime-api/src/asset.rs:65]: TODO??

**Review by fmiguelgarcia — APPROVED** [review]

File: pallets/asset/src/lib.rs [hunk 8]
```diff
@@ -2258,4 +2222,155 @@ impl<T: Trait> Module<T> {
...
+        !Identity::<T>::verify_iu_claims_for_transfer(*ticker, to, from)
+    }
+
+    fn custodian_error(from: PortfolioId, custodian: IdentityId) -> bool {
+        Portfolio::<T>::ensure_portfolio_custody(from, custodian).is_err()
+    }
+
+    fn insufficient_balance(ticker: &Ticker, did: IdentityId, value: T::Balance) -> bool {
+        Self::balance_of(&ticker, did) < value
+    }
```
[the remark this answers is not in the document]
Comment (CJP10) [reply pallets/asset/src/lib.rs:2310]: From my understanding `#[inline]`  It only useful when not use LTO and inlining fns across crates. In this case, all these 1 liners are private and  `#[inline]` would have no impact.

**Comment by maxsam4** [comment]
@CJP10 The schema needs to be updated.

**Review by maxsam4 — APPROVED** [review]

**Pull request merged**